src/levanter/models/mistral.py

```diff
@@ -0,0 +1,606 @@
+    KVHeads = property(lambda self: Axis(name="kv_heads", size=self.num_kv_heads))
+    Layers = property(lambda self: Axis(name="layers", size=self.num_layers))
+    Mlp = property(lambda self: Axis(name="mlp", size=self.intermediate_dim))
+    HeadSize = property(lambda self: Axis(name="head_size", size=self.hidden_dim // self.num_heads))
+
+    def __post_init__(self):
+        assert (
+            self.num_heads % self.num_kv_heads == 0
+        ), f"num_heads={self.num_heads} not divisible by num_kv_heads={self.num_kv_heads}."
+
+    @cached_classproperty
+    def default_hf_checkpoint_converter(cls) -> HFCheckpointConverter["MistralConfig"]:  # type: ignore
+        return HFCheckpointConverter(
+            cls,  # type: ignore
+            "mistralai/Mistral-7B-v0.1",
+            trust_remote_code=True,
+            tokenizer="mistralai/Mistral-7B-v0.1",
+            HfConfigClass=HfMistralConfig,
+        )
+
+    @classmethod
+    def from_hf_config(cls, hf_config: HfConfig):
+        return MistralConfig(
+            seq_len=hf_config.max_position_embeddings,  # this might be too big...
+            hidden_dim=hf_config.hidden_size,
+            intermediate_dim=hf_config.intermediate_size,
+            num_layers=hf_config.num_hidden_layers,
+            num_heads=hf_config.num_attention_heads,
+            num_kv_heads=hf_config.num_key_value_heads,
+            activation_function=hf_config.hidden_act,
+            initializer_range=hf_config.initializer_range,
+            layer_norm_epsilon=hf_config.rms_norm_eps,
+            sliding_window=hf_config.sliding_window,
+        )
+
+    def to_hf_config(self, vocab_size: int, config_overrides: Optional[Dict] = None) -> HfMistralConfig:
+        """Convert to HuggingFace's MistralConfig
+
+        Args:
+            vocab_size (int, optional): Vocabulary size of the tokenizer. Defaults to 32000.
+            config_overrides (dict, optional): Overrides for the config. Defaults to None.
+
+        Returns:
+            HfMistralConfig: HuggingFace's MistralConfig
+        """
+        if config_overrides is None:
+            config_overrides = {}
+
+        return HfMistralConfig(
+            max_position_embeddings=self.seq_len,
+            hidden_size=self.hidden_dim,
+            intermediate_size=self.intermediate_dim,
+            num_hidden_layers=self.num_layers,
+            num_attention_heads=self.num_heads,
+            num_key_value_heads=self.num_kv_heads,
+            hidden_act=self.activation_function,
+            initializer_range=self.initializer_range,
+            rms_norm_eps=self.layer_norm_epsilon,
+            sliding_window=self.sliding_window,
+            vocab_size=vocab_size,
+            **config_overrides,
+        )
+
+    @property
+    def model_type(cls) -> Type["MistralLMHeadModel"]:
+        return MistralLMHeadModel
+
+
+class MistralMlp(eqx.Module, StateDictSerializationMixin):
```

Comment:
so you want to import classes from llama to use directly? I thought you want to make it more like huggingface transformers type of thing so that every model is a separate piece. I can look into reusing llama classes then.

Reply:
Sorry, yeah, it's a delicate balance thing. The goal is to not unduly increase complexity but to be DRY where we can be DRY.

So like, I don't like Transformers' extreme stance of "always repeat yourself" and XTransformers of "one transformer to rule them all" is too extreme too, but something in the middle of "reuse pieces that are the same but don't force unification for the sake of unifying"

So like, we wouldn't want to unify the Gpt2 MLP with the Llama/mistral one, but since the Mistral and Llama ones are literally the same

Reply:
deleted most of the classes. 

I kept the highest module `MistralLMHeadModel` just in case we want to implement some additional logic for attention mask in the future.

I also let `MistralConfig` inherit from `LlamaConfig` to enforce the typing constraints when mistral uses llama modules with its mistral config. Not sure whether `LmConfig.regisster_subclass("mistral")` will cause any trouble in inheritance but so far the tests aren't complaining stuff.